File: src/main/java/com/company/erp/features/users/dto/CreateRoleRequest.java
package com.company.erp.features.users.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;

import java.util.Set;

public record CreateRoleRequest(
        @NotBlank @Size(max = 64)  String code,
        @NotBlank @Size(max = 128) String name,
        @Size(max = 255) String description,
        Set<String> permissions
) {
    public CreateRoleRequest {
        if (permissions == null) permissions = Set.of();
    }
}

File: app/Features/Users/Requests/CreateRoleRequest.php
<?php

namespace App\Features\Users\Requests;

use App\Support\Http\ApiFormRequest;

class CreateRoleRequest extends ApiFormRequest
{
    public function rules(): array
    {
        return [
            'code' => ['required', 'string', 'max:64'],
            'name' => ['required', 'string', 'max:128'],
            'description' => ['nullable', 'string', 'max:255'],
            'permissions' => ['nullable', 'array'],
            'permissions.*' => ['string'],
        ];
    }

    /**
     * @return array<int, string>
     */
    public function permissionCodes(): array
    {
        return array_values(array_unique($this->input('permissions', []) ?: []));
    }
}
